### PR TITLE
Change eval_semantic_segmentation to assume label to be in shape (H, W)

### PR DESCRIPTION
**Pull request by yuyu2172**

### CODE REVIEW

File: chainercv/evaluations/eval_semantic_segmentation.py [hunk 1]
```diff
@@ -59,15 +59,13 @@ def eval_semantic_segmentation(pred_label, gt_label, n_class):
     Args:
         pred_label (array): An integer array of image containing
```
Comment (Hakuyume) [inline chainercv/evaluations/eval_semantic_segmentation.py]: `image` -> `an image`

File: chainercv/evaluations/eval_semantic_segmentation.py [hunk 2]
```diff
@@ -59,15 +59,13 @@ def eval_semantic_segmentation(pred_label, gt_label, n_class):
...
             class labels as values, which is obtained from inference.
-            This has shape :math:`(N, 1, H, W)` or :math:`(1, H, W)`,
+            This has shape :math:`(N, H, W)` or :math:`(H, W)`,
             where :math:`N` is size of the batch, :math:`H` is the height
             and :math:`W` is the width.
```
Comment (Hakuyume) [inline chainercv/evaluations/eval_semantic_segmentation.py]: `H` is the height and W` is the width` -> `H and W are the height and width of the image`

File: chainercv/evaluations/eval_semantic_segmentation.py [hunk 3]
```diff
@@ -59,15 +59,13 @@ def eval_semantic_segmentation(pred_label, gt_label, n_class):
...
         gt_label (array): An integer array of image containing
```
Comment (Hakuyume) [inline chainercv/evaluations/eval_semantic_segmentation.py]: `image` -> `an image`

File: chainercv/evaluations/eval_semantic_segmentation.py (hunk 1 above)
Comment (Hakuyume) [inline chainercv/evaluations/eval_semantic_segmentation.py]: Can we assume that the sizes of input images are same? For example, the sizes of images from `VOCSemanticSegmentationDataset` are different.

**Comment by Hakuyume** [comment]
For the consistency with detection, can we assume all inputs are `numpy.ndarray` (not `cupy.array)?

**Comment by yuyu2172** [comment]
Actually, supporting `cupy.ndarray` is quite useful.
Since reporters take cupy.ndarray as well, I think it is natural to support `cupy.ndrarry`.

I think there is a trade-off between simplicity and utility.

Perhaps, we can change detection code.

File: chainercv/evaluations/eval_semantic_segmentation.py [hunk 4]
```diff
@@ -57,17 +67,16 @@ def eval_semantic_segmentation(pred_label, gt_label, n_class):
...
-            This has shape :math:`(N, 1, H, W)` or :math:`(1, H, W)`,
-            where :math:`N` is size of the batch, :math:`H` is the height
-            and :math:`W` is the width.
-        gt_label (array): An integer array of image containing
-            the ground truth class labels as values. A pixel with value
-            "-1" will be ignored during evaluation. Its shape is similar
-            to :obj:`pred_label`.
-            Its image size is equal to that of :obj:`pred_label`.
-            This should be a one channel CHW formatted image.
+        pred_label (iterable of arrays): A collection of predicted labels.
```
Comment (Hakuyume) [inline chainercv/evaluations/eval_semantic_segmentation.py]: `(iterable of arrays)` -> `(iterable of arrays or array)`
If this function is mainly called with an iterable of arrays, `pred_labels` is better.

File: chainercv/evaluations/eval_semantic_segmentation.py [hunk 5]
```diff
@@ -57,17 +67,16 @@ def eval_semantic_segmentation(pred_label, gt_label, n_class):
...
+            This is an interable of labels or a single label.
+            An array of label has shape :math:`H, W)`.
+            :math:`H` and :math:`W`
+            are height and width of the images. We assume that there are
+            :math:`N` labels.
+        gt_label (iterable of arrays): A collection of the ground truth labels.
+            Its organized similarly to :obj:`pred_label`. A pixel with value
+            "-1" will be ignored during evaluation.
+            We assume that there are :math:`N` labels.
```
Comment (Hakuyume) [inline chainercv/evaluations/eval_semantic_segmentation.py]: Current code can take a single label as `gt_label`. Please add a note.

File: chainercv/evaluations/eval_semantic_segmentation.py [hunk 6]
```diff
@@ -77,30 +86,31 @@ class labels as values, which is obtained from inference.
         the number of images in the input.
```
Comment (Hakuyume) [inline chainercv/evaluations/eval_semantic_segmentation.py]: `MIoU` -> `mIoU`? (I'm not sure)

Comment (Hakuyume) [inline chainercv/evaluations/eval_semantic_segmentation.py]: If the input is a single label, should the outputs be a tuple of scalars?

Comment (Hakuyume) [inline chainercv/evaluations/eval_semantic_segmentation.py]: Or we can simply take only a list of arrays. I think this is not inconvenient because we can easily convert an array to a list of an array.

Comment (yuyu2172) [inline chainercv/evaluations/eval_semantic_segmentation.py]: How about list of arrays or `N, H, W`.
`N, H, W` is probably the most common usage.

Comment (Hakuyume) [inline chainercv/evaluations/eval_semantic_segmentation.py]: `a list of (H, W) or (N, H, W)` looks good to me. In this case, the arguments should be `pred_labels` and `gt_labels`.

Comment (yuyu2172) [inline chainercv/evaluations/eval_semantic_segmentation.py]: Right. Thanks for telling me that.

File: chainercv/evaluations/eval_semantic_segmentation.py [hunk 7]
```diff
@@ -6,18 +6,28 @@
...
+    if isinstance(arrays, xp.ndarray):
+        out_arrays = cuda.to_cpu(arrays)
+    else:
+        out_arrays = []
+        for array in arrays:
+            out_arrays.append(cuda.to_cpu(array))
+    return out_arrays
+
+
+def _fast_hist(pred_labels, gt_labels, n_class):
```
Comment (Hakuyume) [inline chainercv/evaluations/eval_semantic_segmentation.py]: `_fast_hist(pred_label, gt_label, n_class)`

File: chainercv/evaluations/eval_semantic_segmentation.py [hunk 8]
```diff
@@ -57,61 +67,61 @@ def eval_semantic_segmentation(pred_label, gt_label, n_class):
...
-            This has shape :math:`(N, 1, H, W)` or :math:`(1, H, W)`,
-            where :math:`N` is size of the batch, :math:`H` is the height
-            and :math:`W` is the width.
-        gt_label (array): An integer array of image containing
-            the ground truth class labels as values. A pixel with value
-            "-1" will be ignored during evaluation. Its shape is similar
-            to :obj:`pred_label`.
-            Its image size is equal to that of :obj:`pred_label`.
-            This should be a one channel CHW formatted image.
+        pred_labels (iterable of arrays): A collection of predicted
```
Comment (Hakuyume) [inline chainercv/evaluations/eval_semantic_segmentation.py]: `iterable of arrays or array`

Comment (yuyu2172) [inline chainercv/evaluations/eval_semantic_segmentation.py]: Sorry...

File: chainercv/evaluations/eval_semantic_segmentation.py [hunk 9]
```diff
@@ -57,61 +67,61 @@ def eval_semantic_segmentation(pred_label, gt_label, n_class):
...
+            labels. This is a batch of labels whose shape is :math:`(N, H, W)`
+            or a list containing :math:`N` labels. An array of label has shape
+            :math:`H, W`. :math:`H` and :math:`W`
```
Comment (Hakuyume) [inline chainercv/evaluations/eval_semantic_segmentation.py]: `(H, W)`

File: chainercv/evaluations/eval_semantic_segmentation.py (hunk 8 above)
Comment (Hakuyume) [inline chainercv/evaluations/eval_semantic_segmentation.py]: Why did you remove `or array`?

Comment (yuyu2172) [inline chainercv/evaluations/eval_semantic_segmentation.py]: I followed documentation of `predict`.
https://github.com/pfnet/chainercv/blob/master/chainercv/links/model/ssd/ssd.py#L227

Comment (yuyu2172) [inline chainercv/evaluations/eval_semantic_segmentation.py]: But, I guess it works in either way.

Comment (Hakuyume) [inline chainercv/evaluations/eval_semantic_segmentation.py]: I see.

**Comment by Hakuyume** [comment]
LGTM
